## Scraping the UC Davis Course Catalog

The 2022-2023 UC Davis course catalog is online at https://catalog.ucdavis.edu/courses-subject-code/.

In this notebook, I will proceed as follows:
   
       1. Use the lxml package to parse these web pages and then retrieve the names of all subject departments, 
       course titles, and course descriptions. 
       2. Use the extracted information to create a data frame and then clean the data. 
   
I will then visualize this data in Tableau. This will provide a summary of what type of courses are offered at UC Davis and what students would be learning in each specific department.

In [2]:
# import necessary packages

import lxml.html as lx
import re
import requests
import requests_cache
import nltk
import pandas as pd

In [66]:
requests_cache.install_cache('cache')

# write function to retrieve all of the subject names in the catalog (for example: 'Chemistry', 'English', etc.)
def retrieve_subjects(url):
    """
    Input: The URL for the UC Davis course catalog home page.
    Output: A list of all academic subjects in the course catalog.
    """
    # download the web page
    r = requests.get(url)
    doc = r.text

    # parse the web page
    html = lx.fromstring(doc, base_url = url)
    subjects = html.xpath('//ul[@id = "/courses-subject-code/"]/li/a')
    subjects = [x.text_content() for x in subjects]
    
    return subjects

In [67]:
# call function to get course departments/subjects
subjects = retrieve_subjects('https://catalog.ucdavis.edu/courses-subject-code/')

# fix strings in subject list
subjects = [x.replace('&\u200b', '&') for x in subjects]

# get subject codes from subject strings
codes = []
for i in subjects:
    # retrieve the substrings inside the parentheses
    code = i[i.find('(')+1:i.find(')')]
    codes.append(code)

# fix string for 'PBI' subject code    
codes = [x.replace('Graduate Group', 'PBI') for x in codes]

# convert all strings to lower case in order to use them later in URLs
codes = [x.lower() for x in codes]

In [229]:
len(subjects)

216

There are 216 different subjects that are offered at the school.

In [184]:
requests_cache.install_cache('cache')

# write function to retrieve all of the course names and descriptions for each subject
def get_course_info(codes, url):
    """
    Input: A list of subject codes and the url for the UC Davis course catalog.
    Output: The course codes, names, and descriptions.
    """
    titles = []
    descriptions = []
    course_codes = []
    
    # iterate through list of subject codes in order to use the URL for each subject
    for i in codes:
        
        # get URL for current subject
        url_sub = url + i + '/'
        
        # download the web page
        r = requests.get(url_sub)
        doc = r.text
        html = lx.fromstring(doc, base_url = url_sub)
        
        # extract course codes for current subject
        crs_codes = html.xpath('//span[@class = "text courseblockdetail detail-code margin--span text--semibold text--big"]')
        crs_codes = [x.text_content() for x in crs_codes]
        course_codes = course_codes + crs_codes
        
        # extract course titles for current subject
        codetitles = html.xpath('//span[@class = "text courseblockdetail detail-title margin--span text--semibold text--big"]')
        codetitles = [x.text_content() for x in codetitles]
        titles = titles + codetitles
        
        # extract course descriptions for current subject
        code_desc = html.xpath('//p[@class = "courseblockextra noindent"]')
        code_desc = [x.text_content() for x in code_desc]
        descriptions = descriptions + code_desc
        
        
    return course_codes, titles, descriptions

In [185]:
url = 'https://catalog.ucdavis.edu/courses-subject-code/'

# call function to get course codes, titles and descriptions for all courses in catalog
course_codes, titles, descriptions = get_course_info(codes, url)

In [230]:
len(course_codes)

9886

There are currently 9886 different courses offered at the school.

In [191]:
# fix strings
titles = [x.replace('—\xa0', '') for x in titles]
descriptions = [x.replace('Course Description:', '') for x in descriptions]

# create data frame
course_df = pd.DataFrame({'course_code': course_codes, 'course': titles, 'descriptions':descriptions})

# view the first 5 rows of the data frame
course_df.head()

course_code                                            course  \
0     EAE 001     Introduction to Aerospace Science Engineering   
1     EAE 010  From the Wright Brothers to Drones & Quadcopters   
2     EAE 099                  Special Study for Undergraduates   
3     EAE 126          Theoretical & Computational Aerodynamics   
4     EAE 127                     Applied Aircraft Aerodynamics   

                                        descriptions  
0   Description of the field of aerospace enginee...  
1   History of aircraft and its influence on soci...  
2                  Special study for undergraduates.  
3   Development of general equations of fluid mot...  
4   Principles, governing equations, and predicti...

In [199]:
subject_codes = []

# use list of all subjects to get the corresponding subject for each course title
for i in course_codes:
    sub_code = i[0:3]
    for j in subjects:
        if sub_code in j:
            subject_code = j
            subject_codes.append(subject_code)

# add course subjects to the data frame
course_df['subject'] = subject_codes

### Using the Course Catalog Data Frame to Create a Data Frame for my Courses

I have a file with the course codes and names for the courses that I took. I will use the course catalog dataframe to retrieve the course descriptions for the courses that I took.

In [219]:
# open file
my_courses = pd.read_excel('courses.xlsx')
my_courses = my_courses.rename(columns = {'Code' : 'course_code', 'Title' : 'course'})

# use course codes to create data frame with course descriptions from course catalog for my courses
my_courses_df = pd.merge(course_df, my_courses, on = ['course_code'])
my_courses_df

course_code                                           course_x  \
0     BIS 002A  Introduction to Biology: Essentials of Life on...   
1      BIS 010                                   Everyday Biology   
2     CHE 002A                                  General Chemistry   
3     CHE 002B                                  General Chemistry   
4      CMN 003             Interpersonal Communication Competence   
5      ECS 020          Discrete Mathematics For Computer Science   
6     ECS 032A                        Introduction to Programming   
7      ECS 034                 Software Development in UNIX & C++   
8     ECS 036B  Software Development & Object-Oriented Program...   
9     ECS 036C        Data Structures, Algorithms, & Programming    
10     ECS 050  Computer Organization & Machine-Dependent Prog...   
11    ECS 122A                       Algorithm Design & Analysis    
12     ECS 124                Theory & Practice of Bioinformatics   
13     ECS 163              Information Interfaces                  
14     ECS 174                                    Computer Vision   
15     DRA 020                       Introduction to Dramatic Art   
16    ECN 001A                       Principles of Microeconomics   
17    ECN 001B                       Principles of Macroeconomics   
18     ENL 003                         Introduction to Literature   
19    ENL 005F          Introduction to Creative Writing: Fiction   
20    HIS 017A                       History of the United States   
21    HIS 017B                       History of the United States   
22    HIS 111C                                    Ancient History   
23     HDE 012                                    Human Sexuality   
24     LIN 001                        Introduction to Linguistics   
25     MGT 150                              Technology Management   
26    MAT 021A                                           Calculus   
27    MAT 021B                                           Calculus   
28    MAT 021C                                           Calculus   
29    MAT 022A                                     Linear Algebra   
30     MUS 011                                Musics of the World   
31     MUS 126                                     American Music   
32     NAS 012                    Native American/Indigenous Film   
33     SAS 020                                 Genetics & Society   
34     STA 013                              Elementary Statistics   
35     STA 104  Applied Statistical Methods: Nonparametric Sta...   
36     STA 106  Applied Statistical Methods: Analysis of Variance   
37     STA 108   Applied Statistical Methods: Regression Analysis   
38    STA 130B              Mathematical Statistics: Brief Course   
39    STA 131A                 Introduction to Probability Theory   
40     STA 138                       Analysis of Categorical Data   
41    STA 141A           Fundamentals of Statistical Data Science   
42    STA 141B          Data & Web Technologies for Data Analysis   
43     STA 160               Practice in Statistical Data Science   
44    UWP 104T      Writing in the Professions: Technical Writing   

                                         descriptions  \
0    Essentials of life including sources and use ...   
1    Everyday biological concepts using contempora...   
2    Periodic table, stoichiometry, chemical equat...   
3    Continuation of CHE 002A. Thermodynamics, ato...   
4    Communication competence in professional sett...   
5    Discrete mathematics of particular utility to...   
6    Introduction to programming and problem solvi...   
7    UNIX Operating system tools and programming e...   
8    Object-oriented programming in C++. Basic dat...   
9    Design and analysis of data structures for a ...   
10   Comparative study of different hardware archi...   
11   Complexity of algorithms, bounds on complexit...   
12   Fundamental biological, mathematical and algo...   
13   Art and science of inform

## Removing Noise From the Text Descriptions

Common words such as "to" and "of" are not useful for understanding the content of the course descriptions. Before creating word clouds from the course descriptions, I want to remove these words from the 'descriptions' column. 

In [223]:
# remove noise from the text descriptions
nltk.download('stopwords')
from nltk.corpus import stopwords

stopwords = stopwords.words('english')

# remove stopwords from all course descriptions
course_df['descriptions'] = course_df['descriptions'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stopwords)]))
    
my_courses_df['descriptions'] = my_courses_df['descriptions'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stopwords)]))

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Natalie\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


In [243]:
# remove more characters from course descriptions
course_df['descriptions'] = course_df['descriptions'].str.replace('&','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace('&','')
course_df['descriptions'] = course_df['descriptions'].str.replace(',','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace(',','')
course_df['descriptions'] = course_df['descriptions'].str.replace(':','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace(':','')
course_df['descriptions'] = course_df['descriptions'].str.replace('.','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace('.','')
course_df['descriptions'] = course_df['descriptions'].str.replace(';','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace(';','')
course_df['descriptions'] = course_df['descriptions'].str.replace('/','')
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.replace('/','')

# convert all course description text to lower case
course_df['descriptions'] = course_df['descriptions'].str.lower()
my_courses_df['descriptions'] = my_courses_df['descriptions'].str.lower()

In [244]:
# write data frames to csv files

course_df.to_csv('course_catalog.csv')
my_courses_df.to_csv('my_courses.csv')

The Tableau visualization of the course description data can be found at: https://public.tableau.com/app/profile/natalie.perillo/viz/UCDavisCourseCatalog/Dashboard1